src/client_side.cc

```diff
@@ -2810,8 +2743,8 @@ ConnStateData::getSslContextStart()
         Ssl::CertificateProperties certProperties;
         buildSslCertGenerationParams(certProperties);
 
-        // Disable caching for bumpPeekAndSplice mode
-        if (!(sslServerBump && (sslServerBump->act.step1 == Ssl::bumpPeek || sslServerBump->act.step1 == Ssl::bumpStare))) {
+        // Disable caching if the SSL object exists and has attached a CTX object
+        if (!(fd_table[clientConnection->fd].ssl)) {
             sslBumpCertKey.clear();
             Ssl::InRamCertificateDbKey(certProperties, sslBumpCertKey);
             assert(!sslBumpCertKey.isEmpty());
```

Comment:
> 1. What kind of caching are we disabling here? HTTP, TLS sessions, other? Which call inside the if statement will disable that caching? Is the primary purpose of that call to disable caching?

Instead of using a generated certificates cache, we are using a SSL_CTX objects cache.

 In client-first and server-first  bumping mode, when we are entering this code we have not generate yet the OpenSSL required structures to initiate TLS conversation. The required objects are the SSL_CTX (which is also holds server certificate) and then the SSL object (which is built based on SSL_CTX object). In this case we are getting from cache a SSL_CTX object and we are building the SSL object based on this SSL_CTX object

In Peek/Stare modes we are creating SSL_CTX/SSL objects in very early steps and although there is the SSL_set_SSL_CTX method which allow you to set an SSL_CTX object to an SSL object we have consider this procedure unsafe so we decided to regenerate certificates (or get it from helpers cache) and attach the generated certificates to the connection existing (and empty/unconfigured) SSL_CTX object.

(I think it worths to recheck if we can use a cached SSL_CTX object from cache for stare modes.)

> 2. What do we think the if statement conditions mean? The old comment was implying it means we are bumping the client connection (or something like that), but you have removed that so I assume it was wrong.

I changed the old comment in my commit 521148eff. This is because I realized that checks for step1 decisions are not safe and it is better to just check if the SSL/SSL_CTX objects are configured. 

I think the correct comment should be something like the following:
``` // Do not use SSL_CTX objects cache if the SSL/SSL_CTX objects pair for this connection are built ```

Reply:
Please recheck my commit 28c3d4d8 for  the updated comment.

Reply:
> Please recheck my commit 28c3d4d8 for the updated comment.

I appreciate your earlier [explanation](https://github.com/measurement-factory/squid/pull/108#discussion_r629496477). I may be able to polish this code/comments (with your help) further, but I cannot find the 28c3d4d8 commit you mentioned above. Perhaps you have not pushed it to GitHub?